Adds benchmark cli with essential options.

cheqianh opened the pull request:
## Description
This PR added a benchmark module to measure ion-python read/write APIs performance.

The benchmark tool is able to benchmark simpleIon's load/dump and non-blocking APIs with/without C extension enabled. 

## Highlights
* This module uses [`timeit`](https://docs.python.org/3.8/library/timeit.html#timeit.Timer) to benchmark APIs.
* This module uses [`tracealloc`](https://docs.python.org/3/library/tracemalloc.html#tracemalloc.Snapshot.statistics) to profiling memory usage. The memory usage peak is a constant while benchmark time change every execution, so the module profiling test method separately to avoid affecting execution time, see [here](https://github.com/amzn/ion-python/compare/master...cheqianh:benchmark?expand=1#diff-193f9a476a85a550a2aa80e3a9ad54272c2661544bc96455fd6439664e1b9df1R199).
* Warm-ups significantly affect memory usage peak, my assumption is that python needs some memory to set up the first execution and later ones will use cached previously-created objects. So in this tool we use the memory usage peak from the first execution because we are going to find the **peak** of the memory usage for customers.
* Warm-ups significantly affect performance, the first few executions take much longer than later ones. In this tool we uses the average time unlike memory profiling because we want a stable result. But I just think of listing all the fastest time, slowest time and average time is a good idea too.
* Garbage collection time is not measured until we find a reliable way to benchmark it. Python (E.g. C python) uses two strategies to collect garbage (1) reference count and (2) generational garbage collection. In C extension where we worked most closely with C in ion-python, we release reference count of some objects manually which is very hard to benchmark its time so the time we are able to benchmark is more refer to generational garbage collection and as a result we cannot accurately benchmark time used for all kinds of garbage collection. Moreover, disabling garbage collection will affect overall python performance as the free memory is keep being filled but never freed. So we cannot simply use `time spent with garbage collection enabled` minus `time spent with garbage collection disabled` to calculate garbage collection time.
* The tool is able to benchmark [three APIs](https://github.com/amzn/ion-python/compare/master...cheqianh:benchmark?expand=1#diff-d400febd14189603d629816dd332b04576cc206f0f83b943b48f6e5de8a89763R5), simpleIon, iterator and event based non-blocking API. Event based non-blocking API is used by [pure python implementation load/dump](https://github.com/amzn/ion-python/blob/master/amazon/ion/simpleion.py#L307), which is exactly the same thing when we uses `--api simpleion` with disabling [c-extension option](https://github.com/amzn/ion-python/compare/master...cheqianh:benchmark?expand=1#diff-193f9a476a85a550a2aa80e3a9ad54272c2661544bc96455fd6439664e1b9df1R50-R51) by `--c-extension false`. Simpleion module uses iterator for both incremental and DOM mode. The DOM mode uses iterator to go through the whole data and return value. So currently all three APIs can be generated by simpleion API directly (`simpleion` and `iterator `are basically doing the same thing, event based API is just the pure python load/dump). So In this PR, I didn't refactor these APIs into one yet; I'll leave it for now (for example these two methods - [simpleion](https://github.com/amzn/ion-python/compare/master...cheqianh:benchmark?expand=1#diff-193f9a476a85a550a2aa80e3a9ad54272c2661544bc96455fd6439664e1b9df1R143-R161) and [iterator](https://github.com/amzn/ion-python/compare/master...cheqianh:benchmark?expand=1#diff-193f9a476a85a550a2aa80e3a9ad54272c2661544bc96455fd6439664e1b9df1R164-R185)) because two reasons:
(1) They will be different in the future when we start looking into non-blocking API.
(2) Easier debug as they use different options and flags.


## Suggested Review Order
`amazon/ionbenchmark/ion_benchmark_cli.py` is the logic for benchmark cli and [this](https://github.com/amzn/ion-python/compare/master...cheqianh:benchmark?expand=1#diff-193f9a476a85a550a2aa80e3a9ad54272c2661544bc96455fd6439664e1b9df1R356) is its entry point.
`tests/test_benchmark_cli.py` is the uni tests.

## Example Use Case
```
python amazon/ionbenchmark/ion_benchmark_cli.py read --api simpleIon --iterations 100 --warmups 0 test_unit_int
```
This command measures time used for simpleIon API to read `test_unit_int`, it executes 100 times without warming up.
```
python amazon/ionbenchmark/ion_benchmark_cli.py write --api simpleIon --c-extension true test_unit_int
```
This command measures time used for simpleIon to write `test_unit_int` with the C extension enabled. 

Refer to [doc](https://github.com/amzn/ion-python/compare/master...cheqianh:benchmark?expand=1#diff-193f9a476a85a550a2aa80e3a9ad54272c2661544bc96455fd6439664e1b9df1R1-R67) for details.








&nbsp;
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

cheqianh committed:
Adds benchmark cli and its unit tests.

cheqianh commented:
CI/CD failed due to `tabulate` module only support 3.7+, will fix this later.

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -0,0 +1,360 @@
+
+Usage:
+    ion_python_benchmark_cli.py write [--api <api>] [--warmups <int>] [--iterations <int>]
+    [--c-extension <bool>] [--binary <bool>] <input_file>
```
`--binary` should probably take a string, as we will support more formats than just binary and text. For the Java CLI I think we used `--format`. It should also be possible to provide the `--format`, `--api`, and `--c-extension` options more than once to generate a benchmark run that executes more than one possibility.

cheqianh replied:
Addressed

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -0,0 +1,360 @@
+    [--c-extension <bool>] [--binary <bool>] <input_file>
+    ion_python_benchmark_cli.py read [--api <api>] [--warmups <int>] [--iterations <int>]
+    [--c-extension <bool>] <input_file>
+    ion_python_benchmark_cli.py generate
```
We probably won't have this option, as the data generator will be refactored out of the Java CLI and into its own tool. We can remove this and the summary for it below.

cheqianh replied:
Addressed

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -0,0 +1,360 @@
+    ion_python_benchmark_cli.py (-v | --version)
+
+Command:
+    write       Benchmark writing the given input file to the given output format(s). In order to isolate
```
These summaries refer to commands like `--limit` that don't currently exist. That's fine for now, but if we do a release before adding those options, we should remove the references in the summaries.

cheqianh replied:
Removed

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -0,0 +1,360 @@
+Options:
+     -h, --help                          Show this screen.
+
+     --api <api>                        The API to excise (simpleIon, iterator, nonBlocking). `simpleIon` refer to
```
The API enum refers to "event", not "nonBlocking".

cheqianh replied:
Refactor to `event` and `simpleIon`.

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -0,0 +1,360 @@
+                    iterator = ion.load(fp, single_value=single_value, emit_bare_values=emit_bare_values,
+                                        parse_eagerly=parse_eagerly)
+                    data = list(iterator)
+                return data
```
These two branches look like they'll have the same behavior. Is the condition instead supposed to be about `parse_eagerly`?

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -0,0 +1,360 @@
+                    data = list(iterator)
+                return data
+    else:
+        if api == 'simpleIon':
```
Same question as above

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -0,0 +1,360 @@
+    setup_with_gc = generate_simpleion_setup(gc=True, c_ext=c_ext)
+
+    test_code = generate_simpleion_load_test_code(file, emit_bare_values=False, profiling=profiling)
+    test_code_without_wrapper = generate_simpleion_load_test_code(file, emit_bare_values=True, profiling=profiling)
```
Should this be an option, rather than doing it twice?

cheqianh replied:
The tool should be able to support multi-execution and compares benchmark results for different configuration pattens. 

I opened [an issue](https://github.com/amzn/ion-python/issues/226) here, will create a new PR to address this.

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -0,0 +1,360 @@
+        return TOOL_VERSION
+    elif arguments['generate']:
+        print('Generate feature is not supported yet')
+        return 0
```
It won't be included in this tool

cheqianh replied:
Removed.

tgregg commented on amazon/ionbenchmark/API.py:
```diff
@@ -0,0 +1,9 @@
+class API(Enum):
+    """Enumeration of the APIs."""
+    SIMPLE_ION = 'simpleIon'
+    ITERATOR = 'iterator'
```
This should be a command option rather than its own API, since it's controlled by a flag provided to simpleion.

cheqianh replied:
Refactored.

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -0,0 +1,360 @@
+        if not api or api == API.SIMPLE_ION.value:
+            read_micro_benchmark_function = read_micro_benchmark_simpleion
+        elif api == API.ITERATOR.value:
+            read_micro_benchmark_function = read_micro_benchmark_iterator
```
Once ITERATOR is provided via an option, these branches collapse into one, with the iteration option passed into a common method. That allows you to delete some code above.

cheqianh replied:
`iterator` option has the exact same underlying algorithm and behavior as `simpleion`. They use the same API to parse data and gather all of them by `list()`; The C extension [returns iterator](https://github.com/amzn/ion-python/blob/master/amazon/ion/simpleion.py#L493) no matter if iterator options is enabled. I deprecated it.

tgregg replied:
There is a difference in performance characteristics between iterator mode enabled and disabled. For iterator mode, you don't need to hold all top-level values in memory, while when disabled you do. This can make a big difference for large streams. We should definitely have an option to try both modes.

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -0,0 +1,360 @@
+    return file_size, result_with_gc, result_with_raw_value
+
+
+def read_micro_benchmark_iterator(iterations, warmups, c_ext, file, profiling):
```
The `profiling` option seems like it will conflict with the `--profile` option once implemented.

cheqianh replied:
Renamed to `memory_profile`

tgregg commented on tests/test_benchmark_cli.py:
```diff
@@ -0,0 +1,193 @@
+        execution_with_command(['write', file, '--c-extension', 'false'])
+
+    assert file_size == getsize(file) / BYTES_TO_MB
+    # This is a straightforward way to check if C extension is enabled from the top-level python layer.
+    # For most of the case, C extension is supposed to be faster than regular python implementation
+    assert result_with_gc < result_with_gc_2
```
This is potentially brittle. Let's not assert on relative comparisons between result sets in the unit tests.

cheqianh replied:
Removed

tgregg reviewed (commented):
Good start!

When we move this to the amzn repo, we'll need to add the copyright headers to each file.

Can you please paste in an example of output from the tool? Have you done some testing by hand to determine how stable the results are between runs after a reasonable number of warmups and iterations?

cheqianh committed:
refactor --binary to --format.

cheqianh committed:
refactor description and move --limit options to the butoon.

cheqianh committed:
change nonBlocking to event in API description.

cheqianh committed:
Some refactors.

cheqianh committed:
Refactor profiling to memory_profile

cheqianh committed:
deprecated iterator API

cheqianh committed:
fix tests.

cheqianh commented:
Above commits addressed feedback(I responded each comments above), and sample output is in the next comment below. CI/CD failed due to python 3.6, which is deprecated in main repo today. Before I push this PR to the main branch, I'll drop 3.6 and make sure all pipeline passed.

Next step, I'm going to enable benchmark tool to run multiple times with different configuration pattens in one command execution. See issue - https://github.com/amzn/ion-python/issues/226. the issue doesn't block this PR.

cheqianh committed:
copyright

cheqianh commented:
> Can you please paste in an example of output from the tool? Have you done some testing by hand to determine how stable the results are between runs after a reasonable number of warmups and iterations?

Yes, example details below. The data is stable after warm-ups but jumps within a <5% range. Warm-ups significantly impacts the benchmark result.  


## Example output

Here are four examples below, (1) read simple int, (2) write simple int, (3) read `int` vs. `string`, and (4) read `int` vs. `struct`. The `total_time` below in the table indicates how long the parser takes on reading/writing.

#### 1. Simple read command
File - `test_unit_int`:
```
193847912384813276581297346981274618293746812397462318946712389461238384791238481327658129734698127461829374681
```
command:
```
python ion_benchmark_cli.py read --format ion_binary --iterations 10000 --warmups 1000 test_unit_int
```
Output:
I pasted results from two separate executions to show how stable it is. 
```
first time:
╒════════════════╤════════════════╤═════════════╤══════════════════╤════════════════════════╕
│ file_size (MB) │ total_time (s) │ conversion_ │ conversion_time/ │ memory_usage_peak (MB) │
│                │                │ time (s)    │ total_time (%)   │                        │
├────────────────┼────────────────┼─────────────┼──────────────────┼────────────────────────┤
│ 1.06e-04       │ 6.26e-05       │ 1.37e-07    │ 0.22%            │ 4.43e-02               │
╘════════════════╧════════════════╧═════════════╧══════════════════╧════════════════════════╛


second time:
╒════════════════╤════════════════╤═════════════╤══════════════════╤════════════════════════╕
│ file_size (MB) │ total_time (s) │ conversion_ │ conversion_time/ │ memory_usage_peak (MB) │
│                │                │ time (s)    │ total_time (%)   │                        │
├────────────────┼────────────────┼─────────────┼──────────────────┼────────────────────────┤
│ 1.06e-04       │ 6.33e-05       │ 2.67e-07    │ 0.42%            │ 4.43e-02               │
╘════════════════╧════════════════╧═════════════╧══════════════════╧════════════════════════╛


```

#### 2. Simple write command
File - `test_unit_int`:
```
193847912384813276581297346981274618293746812397462318946712389461238384791238481327658129734698127461829374681
```
command:
```
python ion_benchmark_cli.py write --format ion_binary --iterations 10000 --warmups 1000 test_unit_int
```
Output:
Likely, I pasted results from two executions.
```
first execution:
╒════════════════╤════════════════╤════════════════════════╕
│ file_size (MB) │ total_time (s) │ memory_usage_peak (MB) │
├────────────────┼────────────────┼────────────────────────┤
│ 1.06e-04       │ 1.41e-05       │ 5.03e-03               │
╘════════════════╧════════════════╧════════════════════════╛

second execution:
╒════════════════╤════════════════╤════════════════════════╕
│ file_size (MB) │ total_time (s) │ memory_usage_peak (MB) │
├────────────────┼────────────────┼────────────────────────┤
│ 1.06e-04       │ 1.35e-05       │ 5.03e-03               │
╘════════════════╧════════════════╧════════════════════════╛
```

#### 3. Read comparison for simple Ion types (Int vs. string)
File - `test_unit_string` (a string version of test_unit_int)
```
"193847912384813276581297346981274618293746812397462318946712389461238384791238481327658129734698127461829374681"
```
command:
```
python ion_benchmark_cli.py read --format ion_binary --iterations 10000 --warmups 1000 test_unit_string
```
Output:
```
╒════════════════╤════════════════╤═════════════╤══════════════════╤════════════════════════╕
│ file_size (MB) │ total_time (s) │ conversion_ │ conversion_time/ │ memory_usage_peak (MB) │
│                │                │ time (s)    │ total_time (%)   │                        │
├────────────────┼────────────────┼─────────────┼──────────────────┼────────────────────────┤
│ 1.08e-04       │ 5.61e-05       │ 5.48e-07    │ 0.98%            │ 4.43e-02               │
╘════════════════╧════════════════╧═════════════╧══════════════════╧════════════════════════╛

```
Compares the results, **5.61e-5** (parse string) < **6.26e-5** (parse int ). Read string is faster than big integer, I think it's because we parsed `int` to C string first before converting it to python string [here](https://github.com/amzn/ion-python/blob/master/amazon/ion/ioncmodule.c#L1128-L1150) while for `test_unit_str`, parser converts C string to python string directly.

#### 4. Read comparison for complex Ion types (int vs. struct)
File - `test_unit_struct` (a struct version of test_unit_int)
```
{a: 193847912384813276581297346981274618293746812397462318946712389461238384791238481327658129734698127461829374681}
```
command:
```
python ion_benchmark_cli.py read --format ion_binary --iterations 10000 --warmups 1000 test_unit_struct
```
Output:
```
╒════════════════╤════════════════╤═════════════╤══════════════════╤════════════════════════╕
│ file_size (MB) │ total_time (s) │ conversion_ │ conversion_time/ │ memory_usage_peak (MB) │
│                │                │ time (s)    │ total_time (%)   │                        │
├────────────────┼────────────────┼─────────────┼──────────────────┼────────────────────────┤
│ 1.11e-04       │ 7.45e-05       │ 4.82e-06    │ 6.47%            │ 4.47e-02               │
╘════════════════╧════════════════╧═════════════╧══════════════════╧════════════════════════╛
```

Reading `struct` takes 7.45e-5 while reading `int` takes 6.26e-05. it's expected as the extra time is used for constructing an Ion Struct.

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -1,59 +1,51 @@
-     --api <api>                        The API to excise (simpleIon, iterator, nonBlocking). `simpleIon` refer to
-                                        simpleIon's load method. `iterator` refers to simpleIon's iterator type got by
-                                        setting `parse_eagerly` to false. `nonBlocking` refer to ion-python's event
+     --api <api>                        The API to excise (simpleIon, event). `simpleIon` refers to
```
excise -> exercise

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -1,59 +1,51 @@
      -i --iterations <int>              Number of benchmark iterations. [default: 10]
 
-     -c --c-extension <bool>            If enable C extension, note that it only applys to simpleIon module currently.
+     -c --c-extension <bool>            If enables C extension, note that it only applys to simpleIon module currently.
```
If the C extension is enabled

cheqianh committed:
Merge branch 'master' into benchmark

cheqianh committed:
Adds iterator option.

cheqianh commented:
## Description 
The new commit added iterator option and fixed typo. Iterator will go through the whole data stream, DOM API returns all objects directly and stores them in memory. I generated new benchmark data below.  The commits also updated unit tests. 

**File**  
`test_iterator` - repeating `{a: {b: c}}    1  2   3  4` a bunch of times and the file size is 37 KB
```
{a: {b: c}}  1 2 3 4 ... {a: {b: c}}  1 2 3 4 ... {a: {b: c}}  1 2 3 4
```

**Output**
```
python ion_benchmark_cli.py read --warmups 100 --iterations 100 --iterator false test_iterator

╒════════════════╤════════════════╤═════════════╤══════════════════╤════════════════════════╕
│ file_size (MB) │ total_time (s) │ conversion_ │ conversion_time/ │ memory_usage_peak (MB) │
│                │                │ time (s)    │ total_time (%)   │                        │
├────────────────┼────────────────┼─────────────┼──────────────────┼────────────────────────┤
│ 3.49e-02       │ 2.32e-02       │ 0.00e+00    │ 0.00%            │ 4.09e+00               │
╘════════════════╧════════════════╧═════════════╧══════════════════╧════════════════════════╛


python ion_benchmark_cli.py read --warmups 100 --iterations 100 --iterator true test_iterator
╒════════════════╤════════════════╤═════════════╤══════════════════╤════════════════════════╕
│ file_size (MB) │ total_time (s) │ conversion_ │ conversion_time/ │ memory_usage_peak (MB) │
│                │                │ time (s)    │ total_time (%)   │                        │
├────────────────┼────────────────┼─────────────┼──────────────────┼────────────────────────┤
│ 3.49e-02       │ 2.51e-02       │ 0.00e+00    │ 0.00%            │ 4.46e-02               │
╘════════════════╧════════════════╧═════════════╧══════════════════╧════════════════════════╛

```
The execution time of the iterator is a bit longer but uses very little of memory.

cheqianh committed:
temporarily disable pypy.

cheqianh commented:
The CI/CD failed due to incompatible of benchmark tool [tracemalloc](https://docs.python.org/3/library/tracemalloc.html) library and pypy. As discussed offline, we will disable memory profiling for pypy, see https://github.com/amzn/ion-python/issues/227 for details.

Below commit fixed all GHA failures. It did below things:
1. Disabled memory profiling and C extension when `pypy` is used.
2. Removed two redundant tests as it invokes c extension directly which is not allowed with pypy. It should test methods through command-line-style execution helper functions (example [here](https://github.com/cheqianh/ion-python/pull/32/files#diff-18efeaef55437dbe1fc3481d770a8a50aab829eadabfe2c6843ae33addf4a67cR142))
3. Some refactor and rename.

cheqianh committed:
disable memory profiling for pypy

cheqianh commented:
Above commit disabled memory profiling and C extension in `pypy`.

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -0,0 +1,328 @@
+"""A repeatable benchmark tool for ion-python implementation.
+
+Usage:
+    ion_python_benchmark_cli.py write [--api <api>] [--iterator <bool>] [--warmups <int>] [--iterations <int>]
```
The `--iterator` option is only applicable to the `read` command, right?

cheqianh replied:
Thanks for catching up this!

tgregg reviewed (approved):
Nice.

cheqianh committed:
remove iterator option for write

remove iterator for write

fixed a typo
